---
## BWA v0.7.17 [Directory: 3_bwa]: Burrows-Wheeler Alignment (Aligning Sequences to a Reference)
---
#### 1. Create a directory in zsmith10 for `bwa` analyses and change directories to it.
```
cd ../
mkdir 3_bwa
cd 3_bwa
```

#### 2. Create a directory to house the reference genome and all pre-indexed files, then symbolically link them.
```
mkdir solenopsis_genome_index
cd solenopsis_genome_index
ln -s ../../../../raw_data/solenopsis_invicta/genome/UNIL_Sinv_3.0.fasta* .
```

#### 3. Return to the 3_bwa directory and symbolically link trimmed .fastq files from the 2_skewer directory.
```
cd ../
ln -s ../2_skewer/*-trimmed.fastq .
```

#### 4. Load `bwa` v0.7.17 and `samtools` v1.9 using `spack`
```
spack load bwa
spack load samtools@1.9%gcc@8.4.1
```

#### 5. Create a new script file (file extension = .sh) using `nano`. You may need to load `nano` using `spack`.
```
spack load nano
nano bwa.sh
```

#### 6. Here, we will use a for loop to: 1) define basenames for our .fastq files by removing the ".fastq" text and printing to the terminal using echo; 2) loading `bwa` and `samtools` using `spack`; and 3) running `bwa` to align S. invicta sequences to the reference genome, producing SAM files; and 4) piping bwa SAM file output into samtools to convert files into BAM format.
#### In the open nano file, enter the following script:
```
for file in *.fastq
do
    basename=$(echo "$file" | sed 's/.fastq//')
    echo $file
    echo $basename

    spack load bwa
    spack load samtools@1.9%gcc@8.4.1

    bwa mem -t 6 \
    solenopsis_genome_index/UNIL_Sinv_3.0.fasta \
    ${basename}.fastq \
    | samtools view -bSh \
    | samtools sort \
    -@ 3 -m 4G \
    -o ${basename}_sorted.bam
done
```
* the -t flag specified the number of threads for `bwa` to use.
* the -bSh flag in `samtools` converts large SAM files into smaller, binary BAM files with headers.
* the -@ flag designates threads for the job in `samtools`.
* the -m flag designates memory for the job in `samtools`.

#### 7. Run the script, which will produce sorted BAM files, using `bash` with the following command:
```
bash bwa.sh
```

#### 8. To assess number of mapped and supplemental reads, create a script for the following for loop to create read-mapping stats files with `samtools`. Use `bc`'s calculator tool to divide by 4 to correct the number of mapped reads
```
nano samtools_flagstat.sh
```
```
spack load samtools@1.9
spack load bc%gcc@8.4.1

for file in *_sorted.bam
do
        basename=$(echo $file | sed 's/-trimmed_sorted.bam//')
        samtools flagstat $basename-trimmed_sorted.bam > $basename-trimmed_sorted.stats
	
	echo $basename
        echo $(cat $basename-trimmed.fastq | wc -l)/4 | bc

done
```

#### 9. Add read groups to all BAM files for processing in GATK using `java` and `samtools`. First, load `java` (for Picard Tools) and `samtools`. 
```
spack load openjdk@11.0.8_10%gcc@8.4.1
spack load samtools@1.9
```

#### 10. Now, add read groups to each sample and index them with a for loop using Picard Tools.
```
nano picardtools.sh

for f in *_sorted.bam
do
        BASE=$( basename $f | sed 's/_sorted.bam*//g' )
        echo "BASE $BASE"
        
	java -jar /pickett_shared/software/picard-2.27.4/picard.jar \
		AddOrReplaceReadGroups \
		I=${BASE}_sorted.bam \
		O=${BASE}_sorted.RG.bam \
		RGSM=$BASE \
		RGLB=$BASE \
		RGPL=illumina \
		RGPU=$BASE
	samtools index ${BASE}_sorted.RG.bam
done
```  

#### 11. Exit the `nano` text file and run the script using `bash`.
```
bash picardtools.sh
```

#### 12. Double check that BAM files with read groups (.RG.bam files) actually had read groups added by checking the size of both .bam and their corresponding .RG.bam files. Add the flag `-lh` to view files in a list with human-readable sizes.
```
ls -lh *bam
```