---
## 5. GATK v4.1.8.1 : Combining GVCFs and calling SNPs across sample groups
---
#### 1. Return to your personal directory (`/pickett_shared/teaching/EPP622_Fall2022/analysis_test2/zsmith10`), make a new directory for combined `gatk` analyses, and change into it.
```
cd ../
mkdir 5_gatk_combine
cd 5_gatk_combine
```

#### 2. Symbolically link all reference genome files.
```
ln -s /pickett_shared/teaching/EPP622_Fall2022/raw_data/solenopsis_invicta/genome/UNIL_Sinv_3.0.fasta .
ln -s ../4_gatk/UNIL_Sinv_3.0.fasta.fai .
ln -s ../4_gatk/UNIL_Sinv_3.0.dict .
```

#### 3. Link all VCF and GVCF files from the 4_gatk directory.
```
ln -s /pickett_shared/teaching/EPP622_Fall2022/analysis_test2/zsmith10/4_gatk/*vcf .
```

#### 4. Write a script named `gatk_combine_gvcf_create.text.sh` by opening a new file in `nano` and entering the following for loop to generate a script containing the text input required to combine GVCF files using the `gatk` CombineGVCF tool.

* Define your reference as the _Solenopsis invictus_ reference genome using the `-R` flag in `gatk`'s CombineGVCFs tool. 

* Append each echo command to the new script file ( `gatk_combine_gvcf_run.sh` ) using `>>`. This will save the text output into a new, ready-to-run script to generate a combined GVCF. (Warning: extra spaces and tabs will break the script!)

```
nano gatk_combine_gvcf_create.text.sh
```
```
echo "/pickett_shared/software/gatk-4.2.6.1/gatk CombineGVCFs \\
-R UNIL_Sinv_3.0.fasta \\" >> gatk_combine_gvcf_run.sh

for f in *g.vcf
do
        echo "-variant $f \\" >> gatk_combine_gvcf_run.sh
done

echo "-O solenopsis_combined.g.vcf.gz" >> gatk_combine_gvcf_run.sh

```

#### 5. Now that you have created your `gatk_combined_gvcf_create.txt.sh` script, run it using `bash` to automatically create a script with your GVCF input text files and the `gatk` CombineGVCFs tool commands.
```
bash gatk_combine_gvcf_create.text.sh
```

#### 6. Now that your `gatk_combine_gvcf_run.sh` script has been generated. Use `nano` to ensure the new script reads as intended, and run it using `bash` to generate a combined GVCF file using the `gatk` CombineGVCFs tool.
```
bash gatk_combine_gvcf_run.sh
```
*Copy of `gatk_combined_gvcf_create.txt.sh` script output text contained in the secondary `gatk_combine_gvcf_run.sh` script.
```
/pickett_shared/software/gatk-4.2.6.1/gatk CombineGVCFs \
-R UNIL_Sinv_3.0.fasta \
-variant SRR6922141.g.vcf \
-variant SRR6922185.g.vcf \
-variant SRR6922187.g.vcf \
-variant SRR6922236.g.vcf \
-O solenopsis_combined.g.vcf.gz
```

#### 7. Now call variants again using the GenotypeGVCFs tool in `gatk` to generate a combined VCF file.
```
/pickett_shared/software/gatk-4.2.6.1/gatk --java-options "-Xmx10g" GenotypeGVCFs \
   -R UNIL_Sinv_3.0.fasta  \
   -V solenopsis_combined.g.vcf.gz \
   -O solenopsis_combined.vcf.gz
```

#### 8. Check the number of SNPs using `bcftools` to create a `.stats.txt` file. Make sure to load `bcftools` in `spack`.
```
spack load bcftools
bcftools stats solenopsis_combined.vcf.gz > solenopsis_combined.vcf.stats.txt
```
* Variant information in the file output appears so:
```
SN      0       number of samples:      4
SN      0       number of records:      59869
SN      0       number of no-ALTs:      0
SN      0       number of SNPs: 52557
SN      0       number of MNPs: 0
SN      0       number of indels:       7442
SN      0       number of others:       0
SN      0       number of multiallelic sites:   1235
SN      0       number of multiallelic SNP sites:       137
```

#### 9. Before filtering, unzip the `solenopsis_combined.vcf.gz` file in order to view files and ensure filters are working properly along the way.
```
gunzip solenopsis_combined.vcf.gz
```

#### 10. Filter for sequence depth and quality using manual parameters in the `gatk` VariantFiltration tool.
```
/pickett_shared/software/gatk-4.2.6.1/gatk VariantFiltration \
        -R UNIL_Sinv_3.0.fasta  \
        -V solenopsis_combined.vcf \
        -O solenopsis_combined.Basicfilters.vcf \
        -filter-name "basic_filter" -filter "QUAL > 20.0 && DP > 5"
```

#### 11. Check how many SNPs there are in the `Basicfilters` file using `bcftools` to create a `.stats.txt` file. Load `bcftools` using `spack`, if necessary.
```
spack load bcftools
bcftools stats solenopsis_combined.Basicfilters.vcf > solenopsis_combined.Basicfilters.vcf.stats.txt
```

#### 12. Now, filter using the 'gatk' recommended filtering parameters to compare filtering. `gatk` filters help with strand bias, filtering for quality based upon depth, etc.
Note: There is no need to check total SNP counts using `bcftools` because no information is actually removed by filtering, only flagged differently. Thus, output parameters, such as SNP counts, will be identical.

```
/pickett_shared/software/gatk-4.2.6.1/gatk VariantFiltration \
        -R UNIL_Sinv_3.0.fasta  \
        -V solenopsis_combined.vcf \
        -O solenopsis_combined.GATKfilters.vcf \
        -filter-name "QD_filter" -filter "QD < 2.0" \
        -filter-name "FS_filter" -filter "FS > 60.0" \
        -filter-name "MQ_filter" -filter "MQ < 40.0" \
        -filter-name "SOR_filter" -filter "SOR > 4.0" \
        -filter-name "MQRankSum" -filter "MQRankSum < -12.5" \
        -filter-name "ReadPosRankSum" -filter "ReadPosRankSum < -8.0" 
```
#### 13. Gzip the following GVCF files to reduce their storage size:
* solenopsis_combined.vcf 

* solenopsis_combined.Basicfilters.vcf 

* solenopsis_combined.GATKfilters.vcf 
```
gzip solenopsis_combined.vcf
gzip solenopsis_combined.Basicfilters.vcf
gzip solenopsis_combined.GATKfilters.vcf
```

#### 14. Compare filtered and unfiltered files using `bcftools`.
```
bcftools stats -f "PASS,." solenopsis_combined.vcf.gz >solenopsis_combined.vcf.stats.txt
bcftools stats -f "PASS,." solenopsis_combined.Basicfilters.vcf.gz >solenopsis_combined.Basicfilters.vcf.stats.txt
bcftools stats -f "PASS,." solenopsis_combined.GATKfilters.vcf.gz >solenopsis_combined.GATKfilters.vcf.stats.txt
```

#### 15. Search for the number of SNPs in each `.stats.txt` file using `grep`.
```
grep 'number of SNPs:' *stats.txt
```
SNP Output:
```
solenopsis_combined.Basicfilters.vcf.stats.txt:SN       0       number of SNPs: 3194
solenopsis_combined.GATKfilters.vcf.stats.txt:SN        0       number of SNPs: 36851
solenopsis_combined.vcf.stats.txt:SN                    0       number of SNPs: 52557
```

#### 16. Search for the number of indels in each `.stats.txt` file using `grep`.
```
grep 'number of indels:' *stats.txt
```
Indel Output:
```
solenopsis_combined.Basicfilters.vcf.stats.txt:SN       0       number of indels:       333
solenopsis_combined.GATKfilters.vcf.stats.txt:SN        0       number of indels:       5130
solenopsis_combined.vcf.stats.txt:SN                    0       number of indels:       7442
```
